src/boss/azure_worker.go

```diff
@@ -0,0 +1,214 @@
+			worker_i.publicAddr = ""
+		} else {
+			worker_i.publicAddr = *publicWrap.Properties.IPAddress
+		}
+		worker_i.workerId = *conf.Resource_groups.Rgroup[0].Vms[i].Vm.Name
+		worker_i.configPosit = num
+	}
+	parent := &WorkerPool{
+		WorkerPoolPlatform: pool,
+	}
+	return parent
+}
+
+// Is nextId here useful? I store nextId in the pool
+// TODO: maybe store nextId to the config file so that if the boss shut down, it know how to do next time
+func (pool *AzureWorkerPool) NewWorker(workerId string) *Worker {
+	return &Worker{
+		workerId: workerId,
+		workerIp: "",
+	}
+}
+
+// TODO: make AzureCreateVM multiple-threaded
+func (pool *AzureWorkerPool) CreateInstance(worker *Worker) error {
+	log.Printf("creating an azure worker\n")
+	conf, err := AzureCreateVM(worker)
+	if err != nil {
+		return err
+	}
+	var private string
+
+	vmNum := conf.Resource_groups.Rgroup[0].Numvm
+	private = worker.workerIp
+	newDiskName := worker.workerId + "-disk"
+	newNicName := worker.workerId + "-nic"
+	newNsgName := worker.workerId + "-nsg"
+	subnetName := worker.workerId + "-subnet"
+	vnetName := "ol-boss-vnet"
+	publicIPName := ""
+	public := ""
+
+	azworker := &AzureWorker{
+		pool:         pool,
+		workerId:     worker.workerId,
+		configPosit:  vmNum - 1,
+		diskName:     newDiskName,
+		vnetName:     vnetName,
+		nicName:      newNicName,
+		nsgName:      newNsgName,
+		subnetName:   subnetName,
+		publicIPName: publicIPName,
+		privateAddr:  private,
+		publicAddr:   public, // If newly created one, this is ""
+	}
+
+	pool.workerNum += 1
+	pool.nextId = pool.workerNum + 1
+
+	(*pool.workers)[azworker.workerId] = azworker
+	worker.workerId = azworker.workerId
+	worker.workerIp = azworker.privateAddr
+
+	return nil
+}
+
+func (worker *Worker) startWorker() {
+	cwd, err := os.Getwd()
+	if err != nil {
+		panic(err)
+	}
+
+	user, err := user.Current()
+	if err != nil {
+		panic(err)
+	}
+
+	cmd := fmt.Sprintf("cd %s; %s; %s",
+		cwd,
+		"sudo mount -o rw,remount /sys/fs/cgroup",
+		"sudo ./ol worker up -i ol-min -d")
+
+	tries := 10
+	for tries > 0 {
+		sshcmd := exec.Command("ssh", "-i", "~/.ssh/ol-boss_key.pem", user.Username+"@"+worker.workerIp, "-o", "StrictHostKeyChecking=no", "-C", cmd)
+		stdoutStderr, err := sshcmd.CombinedOutput()
+		fmt.Printf("%s\n", stdoutStderr)
+		if err == nil {
+			break
+		}
+		tries -= 1
+		if tries == 0 {
+			fmt.Println(sshcmd.String())
+			panic(err)
+		}
+		time.Sleep(5 * time.Second)
+	}
+}
+
+func (worker *AzureWorker) killWorker() {
+	cwd, err := os.Getwd()
+	if err != nil {
+		panic(err)
+	}
+	user, err := user.Current()
+	if err != nil {
+		panic(err)
+	}
+	cmd := fmt.Sprintf("cd %s; %s", cwd, "sudo ./ol worker down")
+	log.Printf("Try to ssh into the worker and kill the process")
+	tries := 10
+	for tries > 0 {
+		log.Printf("debug: %s\n", worker.privateAddr)
+		sshcmd := exec.Command("ssh", "-i", "~/.ssh/ol-boss_key.pem", user.Username+"@"+worker.privateAddr, "-o", "StrictHostKeyChecking=no", "-C", cmd)
```

Comment:
Should SSH key name be a config option?

Shouldn't the SSH stuff be part of some general code that is shared between the different platforms (GCP, Azure, etc)?